### PR TITLE
[ip-adapter] fix problem using embeds with the plus version of ip adapters

### LINKED ISSUE #7168: How can I get correct ip adapter image embeds? I got 4D tensors and I cannnot use it.
### Describe the bug

IP Adapter image embed should be 3D tensors. But I got 4D tensors.

### Reproduction

```python
import torch
from diffusers import AutoPipelineForText2Image, DDIMScheduler
from diffusers.utils import load_image

pipeline = AutoPipelineForText2Image.from_pretrained(
    "stabilityai/stable-diffusion-xl-base-1.0",
    torch_dtype=torch.float16,
    variant="fp16"
)
pipeline.scheduler = DDIMScheduler.from_config(pipeline.scheduler.config)
pipeline.load_ip_adapter(
    "h94/IP-Adapter",
    subfolder="sdxl_models",
    weight_name=[
        "ip-adapter-plus_sdxl_vit-h.safetensors",
        "ip-adapter-plus-face_sdxl_vit-h.safetensors"
    ] ,
    image_encoder_folder="models/image_encoder"
)
pipeline.set_ip_adapter_scale([0.7, 0.3])
pipeline.enable_model_cpu_offload()

face_image = load_image("https://huggingface.co/datasets/YiYiXu/testing-images/resolve/main/women_input.png")
style_folder = "https://huggingface.co/datasets/YiYiXu/testing-images/resolve/main/style_ziggy"
style_images = [load_image(f"{style_folder}/img{i}.png") for i in range(10)]

image_embeds = pipeline.prepare_ip_adapter_image_embeds(
    ip_adapter_image=[style_images, face_image],
    ip_adapter_image_embeds=None,
    device="cuda",
    num_images_per_prompt=1,
    do_classifier_free_guidance=True
)
torch.save(image_embeds, "image_embeds.ipadpt")

print(f"type: {type(image_embeds)}")
print(f"len: {len(image_embeds)}")
for embeds in image_embeds:
    print(f"shape: {embeds.shape}")
```
outputs is
```
type: <class 'list'>
len: 2
shape: torch.Size([2, 10, 257, 1280])
shape: torch.Size([2, 1, 257, 1280])
```
3D tensors is preferred, but 4D can be obtained. And I cannot use it.
```python
import torch
from diffusers import AutoPipelineForText2Image, DDIMScheduler

pipeline = AutoPipelineForText2Image.from_pretrained(
    "stabilityai/stable-diffusion-xl-base-1.0",
    torch_dtype=torch.float16,
    variant="fp16"
)

pipeline.scheduler = DDIMScheduler.from_config(pipeline.scheduler.config)

pipeline.load_ip_adapter(
    "h94/IP-Adapter",
    subfolder="sdxl_models",
    weight_name=[
        "ip-adapter-plus_sdxl_vit-h.safetensors",
        "ip-adapter-plus-face_sdxl_vit-h.safetensors"
    ],
    image_encoder_folder=None
)
pipeline.set_ip_adapter_scale([0.7, 0.8])

pipeline.to("cuda")

image_embeds_fromfile =  torch.load("image_embeds.ipadpt")

generator = torch.Generator(device="cpu").manual_seed(2024)
image = pipeline(
    prompt="a woman",
    ip_adapter_image_embeds=image_embeds_fromfile,
    negative_prompt="monochrome, lowres, bad anatomy, worst quality, low quality", 
    num_inference_steps=50,
    guidance_scale = 0,
    num_images_per_prompt=1,
    generator=generator,
).images[0]
image.save("result_from_image_embeds.png")
```

### Logs

```shell
ValueError: `ip_adapter_image_embeds` has to be a list of 3D tensors but is 4D
```

### System Info

- `diffusers` version: 0.27.0.dev0
- Platform: Windows-10-10.0.22631-SP0
- Python version: 3.11.6
- PyTorch version (GPU?): 2.2.0+cu118 (True)
- Huggingface_hub version: 0.21.3
- Transformers version: 4.38.1
- Accelerate version: 0.27.2
- xFormers version: not installed
- Using GPU in script?: Yes
- Using distributed or parallel set-up in script?: No

### Who can help?

@sayakpaul
@yiyixuxu

### PR DESCRIPTION
**Pull request by asomoza**
# What does this PR do?
Allows the use of 4D tensors to be able to pass embeds made with the IP Adapter PLUS versions

Fixes #7168

How to test:

```py
import torch

from diffusers import AutoPipelineForText2Image
from diffusers.utils import load_image


pipeline = AutoPipelineForText2Image.from_pretrained(
    "stabilityai/stable-diffusion-xl-base-1.0", torch_dtype=torch.float16, variant="fp16"
)

pipeline.load_ip_adapter(
    "h94/IP-Adapter",
    subfolder="sdxl_models",
    weight_name=[
        "ip-adapter_sdxl_vit-h.safetensors",
        "ip-adapter-plus_sdxl_vit-h.safetensors",
        "ip-adapter-plus-face_sdxl_vit-h.safetensors",
    ],
    image_encoder_folder="models/image_encoder",
)
pipeline.set_ip_adapter_scale([0.1, 0.7, 0.3])
pipeline.to("cuda")

face_image = load_image("https://huggingface.co/datasets/YiYiXu/testing-images/resolve/main/women_input.png")
style_folder = "https://huggingface.co/datasets/YiYiXu/testing-images/resolve/main/style_ziggy"
style_images = [load_image(f"{style_folder}/img{i}.png") for i in range(10)]

prompt = "wonderwoman"
num_images_per_prompt = 1
guidance_scale = 7.5
do_classifier_free_guidance = guidance_scale > 1


with torch.no_grad():
    image_embeds = pipeline.prepare_ip_adapter_image_embeds(
        [face_image, style_images, face_image],
        None,
        "cuda",
        num_images_per_prompt,
        do_classifier_free_guidance,
    )

image = pipeline(
    prompt=prompt,
    ip_adapter_image_embeds=image_embeds,
    negative_prompt="",
    guidance_scale=guidance_scale,
    num_images_per_prompt=num_images_per_prompt,
).images[0]
image.save("result.png")
```

## Who can review?
@yiyixuxu  @sayakpaul

Also cc: @fabiorigano because of [#7186](https://github.com/huggingface/diffusers/pull/7186)

### CODE REVIEW

File: src/diffusers/pipelines/animatediff/pipeline_animatediff.py [hunk 1]
```diff
@@ -404,11 +404,17 @@ def prepare_ip_adapter_image_embeds(
...
                 if do_classifier_free_guidance:
                     single_negative_image_embeds, single_image_embeds = single_image_embeds.chunk(2)
-                    single_negative_image_embeds = single_negative_image_embeds.repeat(num_images_per_prompt, 1, 1)
-                    single_image_embeds = single_image_embeds.repeat(num_images_per_prompt, 1, 1)
+                    single_image_embeds = single_image_embeds.repeat(
+                        num_images_per_prompt, *([1] * len(single_image_embeds.shape[1:]))
+                    )
+                    single_negative_image_embeds = single_negative_image_embeds.repeat(
+                        num_images_per_prompt, *([1] * len(single_negative_image_embeds.shape[1:]))
+                    )
```
Comment (sayakpaul) [inline src/diffusers/pipelines/animatediff/pipeline_animatediff.py:412]: (nit): Can we assign `[1]` in a variable?

**Review by sayakpaul — APPROVED** [review]
Nice! The changes look very clean and simple to me. Thank you!

Should we maybe also add a small note about this support in the IP-Adapter guide? @yiyixuxu WDYT?

**Comment by asomoza** [comment]
Tested all the combinations I could think of with SDXL, the image isn't that nice but is faster to just use all three ip adapters at the same time ^^

Do the SD 1.5 versions have different dimensions? there's six of them.

**Comment by sayakpaul** [comment]
> Do the SD 1.5 versions have different dimensions? there's six of them.

Could do a quick check on the checkpoints maybe?

**Comment by asomoza** [comment]
> Could do a quick check on the checkpoints maybe?

I did and it worked with almost all of them except this one:
[ip-adapter_sd15_vit-G.safetensors](https://huggingface.co/h94/IP-Adapter/blob/main/models/ip-adapter_sd15_vit-G.safetensors)

`RuntimeError: mat1 and mat2 shapes cannot be multiplied (32x1280 and 1024x3072)`

This also reminded me that there's one with the big image encoder for the SDXL ones, so I tested it and it didn't work either:
[ip-adapter_sdxl.safetensors](https://huggingface.co/h94/IP-Adapter/blob/main/sdxl_models/ip-adapter_sdxl.safetensors)

`RuntimeError: mat1 and mat2 shapes cannot be multiplied (82240x1664 and 1280x1280)`

but those errors are not related to this PR

**Comment by sayakpaul** [comment]
> This also reminded me that there's one with the big image encoder for the SDXL ones, so I tested it and it didn't work either:
[ip-adapter_sdxl.safetensors](https://huggingface.co/h94/IP-Adapter/blob/main/sdxl_models/ip-adapter_sdxl.safetensors)

I see. I think that needs fixing then. Would you mind opening an issue for this and we can work on that in a separate PR?

**Comment by asomoza** [comment]
> I see. I think that needs fixing then. Would you mind opening an issue for this and we can work on that in a separate PR?

It was an obvious mistake on my part, since both of those use a different image encoder and I was using them in combination with the normal ones, that was the error, I can't mix the adapters that use different image encoders since we load one for all of them.

They work as expected if I use them alone.

**Review by yiyixuxu — APPROVED** [review]
thank you!

**Pull request merged**